crates/cheatable-starknet/src/rpc.rs

```diff
@@ -489,6 +564,75 @@ impl CheatableSyscallHandler<'_> {
     }
 }
 
+#[derive(Debug)]
+pub struct SingleSegmentResponse {
+    segment: ReadOnlySegment,
+}
+
+impl SyscallResponse for SingleSegmentResponse {
+    fn write(self, vm: &mut VirtualMachine, ptr: &mut Relocatable) -> WriteResponseResult {
+        write_segment(vm, ptr, self.segment)
+    }
+}
+
+pub fn call_contract_syscall(
+    request: CallContractRequest,
+    vm: &mut VirtualMachine,
+    syscall_handler: &mut CheatableSyscallHandler<'_>,
+    remaining_gas: &mut u64,
+) -> SyscallResult<SingleSegmentResponse> {
+    let storage_address = request.contract_address;
+    let mut entry_point = CallEntryPoint {
+        class_hash: None,
+        code_address: Some(storage_address),
+        entry_point_type: EntryPointType::External,
+        entry_point_selector: request.function_selector,
+        calldata: request.calldata,
+        storage_address,
+        caller_address: syscall_handler.syscall_handler.storage_address(),
+        call_type: CallType::Call,
+        initial_gas: *remaining_gas,
+    };
+    let retdata_segment = execute_inner_call(&mut entry_point, vm, syscall_handler, remaining_gas)?;
+
+    Ok(SingleSegmentResponse {
+        segment: retdata_segment,
+    })
+}
+
+pub fn execute_inner_call(
+    call: &mut CallEntryPoint,
+    vm: &mut VirtualMachine,
+    syscall_handler: &mut CheatableSyscallHandler<'_>,
+    remaining_gas: &mut u64,
+) -> SyscallResult<ReadOnlySegment> {
+    let call_info = execute_call_entry_point(
+        call,
+        syscall_handler.syscall_handler.state,
+        syscall_handler.cheated_state,
+        syscall_handler.syscall_handler.resources,
+        syscall_handler.syscall_handler.context,
+    )?;
+
+    let raw_retdata = &call_info.execution.retdata.0;
+
+    if call_info.execution.failed {
+        // TODO(spapini): Append an error word according to starknet spec if needed.
+        // Something like "EXECUTION_ERROR".
+        return Err(SyscallExecutionError::SyscallError {
```

Comment:
Is it copied from cairo? If so can we mark what parts are modified by us? Like this:
https://github.com/foundry-rs/starknet-foundry/pull/309#discussion_r1276194005

Reply:
Added some extra documentation